include/deal.II/lac/petsc_ts.h

```diff
@@ -0,0 +1,706 @@
+                                  nullptr));
+          SNES snes;
+          AssertTS(TSGetSNES(ts, &snes));
+          if (!A)
+            AssertTS(SNESSetUseMatrixFree(snes, PETSC_TRUE, PETSC_FALSE));
+
+          if (!P)
+            {
+              DM dm;
+              AssertTS(SNESGetDM(snes, &dm));
+              AssertTS(DMSetMatType(dm, MATSHELL));
+            }
+        }
+      else
+        {
+          if (explicit_jacobian)
+            {
+              AssertTS(TSSetRHSJacobian(ts,
+                                        A ? A->petsc_matrix() :
+                                            (P ? P->petsc_matrix() : nullptr),
+                                        P ? P->petsc_matrix() : nullptr,
+                                        _dealii_ts_rhsjacobian_,
+                                        nullptr));
+            }
+
+          if (implicit_jacobian)
+            {
+              AssertTS(TSSetIJacobian(ts,
+                                      A ? A->petsc_matrix() :
+                                          (P ? P->petsc_matrix() : nullptr),
+                                      P ? P->petsc_matrix() : nullptr,
+                                      _dealii_ts_ijacobian_,
+                                      nullptr));
+            }
+        }
+
+
+      // In case solve_for_jacobian_system is provided, create a shell
+      // preconditioner wrapping the user call. The internal Krylov
+      // solver will apply the preconditioner only once. This choice
+      // can be overriden by command line and users can use any other
+      // Krylov method if their solve is not accurate enough.
+      PreconditionShell precond(
+        PetscObjectComm(reinterpret_cast<PetscObject>(ts)));
+      if (solve_for_jacobian_system)
+        {
+          precond.vmult = [&](VectorBase &      indst,
+                              const VectorBase &insrc) -> int {
+            VectorType       dst(static_cast<const Vec &>(indst));
+            const VectorType src(static_cast<const Vec &>(insrc));
+            return solve_for_jacobian_system(src, dst);
+          };
+
+          SNES snes;
+          KSP  ksp;
+          AssertTS(TSGetSNES(ts, &snes));
+          AssertTS(SNESGetKSP(snes, &ksp));
+          AssertTS(KSPSetType(ksp, KSPPREONLY));
+          AssertTS(KSPSetPC(ksp, precond.get_pc()));
+        }
+
+      // Attach user monitoring routine
+      if (monitor)
+        AssertTS(TSMonitorSet(ts, _dealii_ts_monitor_, nullptr, nullptr));
+
+      AssertTS(TSSetFromOptions(ts));
+
+      AssertTS(TSSolve(ts, nullptr));
+
+      PetscInt nt;
+      AssertTS(TSGetStepNumber(ts, &nt));
+      return nt;
+    }
+
+    /**
+     * Integrate the differential-algebraic equations.
+     * Here we also pass the matrices to handle Jacobians.
+     */
+    unsigned int
+    solve(VectorType &solution, PMatrixType &P)
+    {
+      reinit_matrices(P);
+      return solve(solution);
+    }
```

Comment:
The function only gets one matrix, but it is unclear to me what it corresponds to and where it is used.

Reply:
It should be clearer from the documentation now.